# Исследование надежности заемщиков


## Откроем таблицу и изучите общую информацию о данных

**Задание 1. Импортируем библиотеку pandas. Считаем данные из csv-файла в датафрейм и сохраним в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')


**Задание 2. Выведем первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведем основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удалим пропуски

**Задание 4. Выведем количество пропущенных значений для каждого столбца. Используем комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаем на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработаем значения в этом столбце: заменим все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведем медианное значение трудового стажа `days_employed` в днях.**

In [7]:

data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведем перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведем перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполним пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедимся, что все пропуски заполнены. Проверим себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Заменим вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработаем неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру. Проверим остальные столбцы.**

In [14]:

data['education'] = data['education'].str.lower()

**Задание 15. Выведем на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалим их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используем собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:



data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведем на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используем собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучим данные в столбце `purpose` и определим, какие подстроки помогут  правильно определить категорию.**

In [20]:



def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:

data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуем данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# функция, которая посчитает нам отношение в процентах
def make_proportion(pdSerises):
    return str(round((pdSerises.sum() / pdSerises.count()) * 100, 3)) + '%'

In [23]:
def common(data_pivot):
    data_temp = data_pivot.rename(columns={'sum':'Всего кредитополучателей', 'count':'Всего должников', 'mean':'Доля должников'})
    return(data_temp)

In [24]:
data_pivot = common(data.pivot_table(index=['children'], values='debt', aggfunc=['sum', 'count', 'mean']))
#print(data_temp)
data_pivot

Всего кредитополучателей Всего должников Доля должников
                             debt            debt           debt
children                                                        
0                            1063           14091       0.075438
1                             444            4808       0.092346
2                             194            2052       0.094542
3                              27             330       0.081818
4                               4              41       0.097561
5                               0               9       0.000000

# **Вывод:** 

Разница колеблится в пределах 1%, что не выводит явной закономерности, но меньше всего вероятность у кредиторов без детей, а по кредиторам с 5ю детьми выборка слишком мала, чтобы отображалась объективная статистика

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
data_pivot = data.pivot_table(index=['family_status'], values='debt', aggfunc=['sum', 'count', 'mean'])
print(data_pivot)

                       sum  count      mean
                      debt   debt      debt
family_status                              
Не женат / не замужем  273   2796  0.097639
в разводе               84   1189  0.070648
вдовец / вдова          63    951  0.066246
гражданский брак       385   4134  0.093130
женат / замужем        927  12261  0.075606


#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
# Создадим функцию, которая распределит широкий диапазон доходов в несколько категорий.
def make_income_category(total_income):
    if total_income < 50000:
        return 'До 50000'
    if 50000 <= total_income < 100000:
        return 'От 50 до 100'
    if 100000 <= total_income < 150000:
        return 'От 100 до 150'
    if 150000 <= total_income < 200000:
        return 'От 150 до 200'
    if 200000 <= total_income < 500000:
        return 'От 200 до 500'
    return 'Больше 500'


data['income_category'] = data['total_income'].apply(make_income_category)

In [27]:
# соберем сводную таблицу. Сформируем данные по income_category для колонки debt. Применим функцию для значений sum и count.
data_pivot = data.pivot_table(index=['income_category'], values=["debt"], aggfunc=['sum', 'count', make_proportion])
#print(data_temp)
data_pivot

sum count make_proportion
                debt  debt            debt
income_category                           
Больше 500        14   222          6.306%
До 50000          23   371          6.199%
От 100 до 150    619  7110          8.706%
От 150 до 200    403  4738          8.506%
От 200 до 500    342  4817            7.1%
От 50 до 100     331  4073          8.127%

**Вывод:** 

Чем больше уровень дохода, тем ниже необходимость в потребительских кредитах и ниже процент просрочки платежа

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [28]:
data_pivot = data.pivot_table(index=['purpose_category'], values='debt', aggfunc=['sum', 'count', 'mean'])
print(data_pivot)

                          sum  count      mean
                         debt   debt      debt
purpose_category                              
операции с автомобилем    400   4279  0.093480
операции с недвижимостью  780  10751  0.072551
получение образования     369   3988  0.092528
проведение свадьбы        183   2313  0.079118


**Вывод:** 

В нише свадьбы и операции с недвижимостью наблюдается меньше просрочек, чем в образовании и автомобилях

#### 3.5 Приведем возможные причины появления пропусков в исходных данных.

Можно предположить, что корректность данных в ячейках, если не брать технические сбои и баги, зависит от того обязательны ли поля к заполнению, есть ли дополнительная проверка роботами введенных данных на наличие отклонения их от стандартных вариантов, которые должны быть обусловленны програмистами не давая возможности пользователю отклонится от требуетых параметров.

#### 3.6 Объясним, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медианные заначения более реалестично отаржают 'среднюю температуру по больнице' так как срезают всплески, исключаяя вывод средней зарплаты в 500тыс.рублей на выборке, где есть зарплаты в 5тр и 1 000 000 000 рублей.

### Шаг 4: общий вывод.

<h1>Детализация</h1>
<div class="alert alert-info" role="alert" >

 Самый привлекательный вид заемщике - это человек без детей,состоявший в официальных отношениях, имеющий высокий уровень дохода и берущий деньги на недвижимость.  
    
  "Разброс" просроченных платежей по каждой из категорий приблезительно одинаковый. В среднем, это диапазоны от 7 до 10 процентов. В остальном разница  не сильно отличается. Однако если смотреть в относительных величинах разница составляет в среднем 2 процента.

Детально по категориям:

- Влияние наличия детей на вероятность просрочки платежа
- По наличию детей
 1. 5	детей	 - 0.000000 в расчет не берется
 2. 0	детей	 - 0.075438
 3. 3	детей	 - 0.081818
 4. 1	ребенок  - 0.092346
 5. 2	детей    - 0.094542
 6. 4	детей	 - 0.097561
- Абсолютная разница между минимальным и максимальным значением - 2.2%
- Влияние семейного положения на вероятность просрочки платежа
- По семейному положению
 1. Не женат / не замужем  - 0.097639
 2. гражданский брак	   - 0.093130
 3. женат / замужем	- 0.075606
 4. в разводе	- 0.070648
 5. вдовец / вдова	- 0.066246
- Абсолютная разница между минимальным и максимальным значением - 3.15%
- Влияние уровня дохода на просрочку платежа
- По уровню дохода
 1. От 100 до 150 - 8.706%
 2. От 150 до 200 - 8.506%
 3. От 50 до 100 - 8.127%
 4. От 200 до 500 - 7.1%
 5. Больше 500 - 6.306%
 6. До 50000	- 6.199%
- Абсолютная разница между минимальным и максимальным значением - 2.5%
- Влияние цели кредита на вероятность возврата кредита в срок
По цели кредита
 1. операции с автомобилем - 9.348%
 2. получение образования - 9.253%
 3. проведение свадьбы	- 7.912%
 4. операции с недвижимостью	- 7.255%
- Абсолютная разница между минимальным и максимальным значением - 2.09%
    Таким образом, несмотря на схожий в процентном отношении разброс целевых значений, наибольшее влияние на вероятность просрочки показывает именно семейное положение. Во вторую очередь, следует обращать внимание на уровень дохода. Наименьшее влияние оказывают цели кредита и наличие детей.
</div>